runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -0,0 +1,642 @@
+                        ? (V) CorfuTable.UndoNullable.NULL
+                        : previousState.get(k))));
+        return builder.build();
+    }
+
+    /** Undo a remove, given the current state of the map, an undo record
+     * and the arguments to the remove command to undo.
+     *
+     * @param table           The state of the map after the put to undo
+     * @param undoRecord    The undo record generated by undoRemoveRecord
+     */
+    @DontInstrument
+     void undoPutAll(CorfuTable<K, V, F, I> table, Map<K,V> undoRecord,
+                            Map<? extends K, ? extends V> m) {
+        undoRecord.entrySet().forEach(e -> {
+                    if (e.getValue() == CorfuTable.UndoNullable.NULL) {
+                        undoRemove(table, null, e.getKey());
+                    } else {
+                        undoRemove(table, e.getValue(), e.getKey());
+                    }
+                }
+        );
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Mutator(name = "put", noUpcall = true)
+    public void insert(@ConflictParameter K key, V value) {
+        V previous = mainMap.put(key, value);
+        // If we have index functions, update the secondary indexes.
+        if (indexFunctions.size() > 0) {
+            unmapSecondaryIndexes(key, previous);
+            mapSecondaryIndexes(key, value);
+        }
+    }
+
+    /**
+     * Returns a filtered {@link List} view of the values contained in this map.
+     * This method has a memory/CPU advantage over the map iterators as no deep copy
+     * is actually performed.
+     *
+     * @param p java predicate (function to evaluate)
+     * @return a view of the values contained in this map meeting the predicate condition.
+     */
+    @Accessor
+    public List<V> scanAndFilter(Predicate<? super V> p) {
+        return mainMap.values().parallelStream().filter(p).collect(Collectors.toList());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public Collection<Map.Entry<K, V>> scanAndFilterByEntry(Predicate<? super Map.Entry<K, V>>
+                                                                    entryPredicate) {
+        return mainMap.entrySet().parallelStream().filter(entryPredicate).collect(Collectors
+                .toCollection(ArrayList::new));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @MutatorAccessor(name = "remove", undoFunction = "undoRemove",
+                                undoRecordFunction = "undoRemoveRecord")
+    @SuppressWarnings("unchecked")
+    public V remove(@ConflictParameter Object key) {
+        V previous =  mainMap.remove(key);
+        unmapSecondaryIndexes((K) key, previous);
+        return previous;
+    }
+
+    @DontInstrument
+    protected V undoRemoveRecord(CorfuTable<K, V, F, I> table, K key) {
+        return table.mainMap.get(key);
+    }
+
+    @DontInstrument
+    protected void undoRemove(CorfuTable<K, V, F, I> table, V undoRecord, K key) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)